pkg/pgmodel/querier/series_exemplar.go

```diff
@@ -0,0 +1,186 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package querier
+
+import (
+	"context"
+	"fmt"
+	"github.com/timescale/promscale/pkg/pgmodel/cache"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgmodel/common/schema"
+	"github.com/timescale/promscale/pkg/pgmodel/model"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+const getExemplarLabelPositions = "SELECT * FROM " + schema.Catalog + ".get_exemplar_label_key_positions($1::TEXT)"
+
+type exemplarSeriesRow struct {
+	metricName string
+	labelIds   []int64
+	data       []exemplarRow
+}
+
+type exemplarRow struct {
+	time        time.Time
+	value       float64
+	labelValues []string // Exemplar label values.
+}
+
+// appendExemplarRows adds new results rows to already existing result rows and
+// returns the as a result.
+func appendExemplarRows(metricName string, in pgxconn.PgxRows) (rows []exemplarSeriesRow, err error) {
+	if in.Err() != nil {
+		return rows, in.Err()
+	}
+	seriesRowMap := make(map[string]*exemplarSeriesRow)
+	// Note: The rows of multiple exemplar query, are such that `rows` here contains different series, represented by
+	// different label_ids array, that satisfy the given labels.
+	for in.Next() {
+		var (
+			err      error
+			row      exemplarRow
+			labelIds []int64
+		)
+		err = in.Scan(&labelIds, &row.time, &row.value, &row.labelValues)
+		if err != nil {
+			log.Error("err", err)
+			return rows, err
+		}
+
+		key := fmt.Sprintf("%v", labelIds)
+		if existingSeriesRow, exists := seriesRowMap[key]; exists {
+			existingSeriesRow.data = append(existingSeriesRow.data, row)
+			continue
+		}
+		// New exemplar series.
+		seriesRow := &exemplarSeriesRow{
+			metricName: metricName,
+			labelIds:   labelIds,
+			data:       make([]exemplarRow, 0, 1),
+		}
+		seriesRow.data = append(seriesRow.data, row)
+		seriesRowMap[key] = seriesRow
+	}
+	return getExemplarSeriesRows(seriesRowMap), in.Err()
+}
+
+func getExemplarSeriesRows(m map[string]*exemplarSeriesRow) []exemplarSeriesRow {
+	s := make([]exemplarSeriesRow, len(m))
+	i := 0
+	for _, v := range m {
+		s[i] = *v
+		i++
+	}
+	return s
+}
```

Comment:
I would inline this into the function above

Reply:
I think its helps in reading. So keeping as it is unless a hard requirement.